Mathlib/Order/KonigLemma.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2024 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import Mathlib.Order.Atoms.Finite
+import Mathlib.Order.Grade
+
+/-!
+# Kőnig's infinity lemma
+
+Kőnig's infinity lemma is most often stated as a graph theory result :
+every infinite, locally finite connected graph contains an infinite path.
+It has links to computability and proof theory, and is in fact a specialization
+of `nonempty_sections_of_finite_cofiltered_system`.
+It is stated and proved here with fewer dependencies.
+
+In practice, most applications are not to an abstract graph,
+but to a concrete collection of objects that are organized in a graph-like way,
+often where the graph is a rooted tree representing a graded order.
+In fact, the lemma is in fact most easily stated and proved
+in terms of covers in a strongly atomic order rather than a graph;
+in this setting, the proof is almost trivial.
+
+We leave the explicitly graph-theoretic version of the statement as TODO.
+
+## Main Results
+
+* `exists_seq_covby_of_forall_covby_finite` : Kőnig's lemma for strongly atomic orders.
+
+* `exists_orderEmbedding_covby_of_forall_covby_finite` : Kőnig's lemma, where the sequence
+  is given as an `OrderEmbedding` instead of a function.
+
+* `exists_orderEmbedding_covby_of_forall_covby_finite_of_bot` : Kőnig's lemma where the sequence
+  starts at the minimum of an infinite type.
+
+* `exist_seq_forall_proj_of_forall_finite` : Kőnig's lemma applied to an order on a sigma-type
+  `(i : ℕ) × α i`.
```

Comment:
I took a pass at this:
```text
Kőnig's infinity lemma is most often stated as a graph theory result:
every infinite, locally finite connected graph contains an infinite path.
It has links to computability and proof theory, and it has a number of formulations.

In practice, most applications are not to an abstract graph,
but to a concrete collection of objects that are organized in a graph-like way,
often where the graph is a rooted tree representing a graded order.
In fact, the lemma is in fact most easily stated and proved
in terms of covers in a strongly atomic order rather than a graph;
in this setting, the proof is almost trivial.

A common formulation of Kőnig's lemma is in terms of directed systems,
with the grading explicitly represented using an `ℕ`-indexed family of types,
which we also provide in this module.
This is a specialization of the much more general `nonempty_sections_of_finite_cofiltered_system`,
which goes through topology and category theory,
but it is stated and proved independently with much fewer dependencies.

We leave the explicitly graph-theoretic version of the statement as TODO.

## Main Results

* `exists_seq_covby_of_forall_covby_finite` : Kőnig's lemma for strongly atomic orders.

* `exists_orderEmbedding_covby_of_forall_covby_finite` : Kőnig's lemma, where the sequence
  is given as an `OrderEmbedding` instead of a function.

* `exists_orderEmbedding_covby_of_forall_covby_finite_of_bot` : Kőnig's lemma where the sequence
  starts at the minimum of an infinite type.

* `exist_seq_forall_proj_of_forall_finite` : Kőnig's lemma for inverse systems,
  proved using the above applied to an order on a sigma-type `(i : ℕ) × α i`.
```

Reply:
Thanks!